File: Python/132_Palindrome_Partitioning_II.py
class Solution:
    def minCut(self, s: str) -> int:
        dp = [i - 1 for i in range(0, len(s) + 1)]
        for i in range(0, len(s)):
            j = 0
            while i + j < len(s) and i - j >= 0 and s[i + j] == s[i - j]:
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j] + 1)
                j += 1

            j = 0
            while i + j < len(s) and i - j - 1 >= 0 and s[i + j] == s[i - j - 1]:
                dp[i + j + 1] = min(dp[i + j + 1], dp[i - j - 1] + 1)
                j += 1

        return dp[-1]

File: golang/132_Palindrome_Partitioning_II.go
func minCut(s string) int {
	n := len(s)
	var dp = make([]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = i - 1
	}
	for i := 0; i < n; i++ {

		for j := 0; (i+j < n) && (i-j >= 0) && (s[i+j] == s[i-j]); j++ {
			dp[i+j+1] = min(dp[i+j+1], dp[i-j]+1)
		}
		for j := 0; (i+j < n) && (i-j-1 >= 0) && (s[i+j] == s[i-j-1]); j++ {
			dp[i+j+1] = min(dp[i+j+1], dp[i-j-1]+1)
		}
	}
	return dp[n]
}

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}
